client/my-sites/people/role-select/index.jsx

```diff
@@ -2,138 +2,83 @@
-export default React.createClass( {
-	displayName: 'RoleSelect',
-
-	getInitialState() {
-		return ( {
-			roles: this.props.siteId ? RolesStore.getRoles( this.props.siteId ) : {}
-		} );
-	},
-
-	componentWillMount() {
-		if ( this.props.includeFollower ) {
-			this.refreshRoles();
-		}
-	},
-
-	componentDidMount() {
-		RolesStore.on( 'change', this.refreshRoles );
-		this.fetchRoles();
-	},
-
-	componentWillUnmount() {
-		RolesStore.removeListener( 'change', this.refreshRoles );
-	},
-
-	componentWillReceiveProps( nextProps ) {
-		const siteId = nextProps.siteId || this.props.siteId;
-		this.fetchRoles( siteId );
-		this.refreshRoles( nextProps );
-	},
-
-	getWPCOMFollowerRole( siteId ) {
-		siteId = siteId ? siteId : this.props.siteId;
-
-		const site = sites.getSite( siteId ),
-			displayName = site.is_private
-			? this.translate( 'Viewer', { context: 'Role that is displayed in a select' } )
-			: this.translate( 'Follower', { context: 'Role that is displayed in a select' } );
-
-		return {
-			follower: {
-				display_name: displayName,
-				name: 'follower'
-			}
-		};
-	},
-
-	refreshRoles( nextProps ) {
-		const siteId = nextProps && nextProps.siteId ? nextProps.siteId : this.props.siteId;
-
-		if ( siteId ) {
-			let siteRoles = RolesStore.getRoles( siteId );
-
-			if ( this.props.includeFollower ) {
-				siteRoles = Object.assign( {}, this.getWPCOMFollowerRole(), siteRoles );
-			}
-
-			this.setState( {
-				roles: siteRoles
-			} );
-		}
-	},
-
-	fetchRoles( siteId = this.props.siteId ) {
-		if ( ! siteId ) {
-			debug( 'siteId not set' );
-			return;
-		}
-
-		if ( Object.keys( RolesStore.getRoles( siteId ) ).length ) {
-			debug( 'initial fetch not necessary' );
-			return;
-		}
-
-		debug( 'Fetching roles for ' + siteId );
-
-		// defer fetch requests to avoid dispatcher conflicts
-		setTimeout( function() {
-			const fetching = RolesStore.isFetching( siteId );
-			if ( fetching ) {
-				return;
-			}
-			RolesActions.fetch( siteId );
-		}, 0 );
-	},
+import QuerySites from 'components/data/query-sites';
+import QuerySiteRoles from 'components/data/query-site-roles';
+import { getSite } from 'state/sites/selectors';
+import { getSiteRoles } from 'state/site-roles/selectors';
+
+const getWpcomFollowerRole = ( { site, translate } ) => {
+	const displayName = site.is_private
+		? translate( 'Viewer', { context: 'Role that is displayed in a select' } )
+		: translate( 'Follower', { context: 'Role that is displayed in a select' } );
+
+	return {
+		display_name: displayName,
+		name: 'follower'
+	};
+};
+
+const RoleSelect = ( props ) => {
+	let { siteRoles } = props;
+	const { site, includeFollower, siteId, id, explanation, translate } = props;
+	const omitProps = [
+		'site',
+		'key',
+		'siteId',
+		'includeFollower',
+		'explanation',
+		'siteRoles',
+		'dispatch',
+		'moment',
+		'numberFormat',
+		'translate'
+	];
+
+	if ( site && siteRoles && includeFollower ) {
+		siteRoles = siteRoles.concat( getWpcomFollowerRole( props ) );
```

Comment:
Do you think this is something that belongs in a state selector? In other words, should `getSiteRoles` include the follower role automatically when the site is not private?